dipy/reconst/ivim.py

```diff
@@ -0,0 +1,463 @@
+from dipy.reconst.base import ReconstModel
+from dipy.reconst.dti import _min_positive_signal, apparent_diffusion_coef
+from dipy.reconst.dti import TensorModel, mean_diffusivity
+from dipy.reconst.vec_val_sum import vec_val_vect
+from dipy.core.sphere import Sphere
+
+from distutils.version import LooseVersion
+import scipy
+
+SCIPY_LESS_0_17 = LooseVersion(scipy.version.short_version) < '0.17'
+
+if not SCIPY_LESS_0_17:
+    least_squares = scipy.optimize.least_squares
+else:
+    leastsq = scipy.optimize.leastsq
+
+
+def ivim_function(params, bvals):
+    """The Intravoxel incoherent motion (IVIM) model function.
+
+    The IVIM model assumes that biological tissue includes a volume fraction
+    f of water flowing in perfused capillaries, with a perfusion
+    coefficient D* and a fraction (1-f) of static (diffusion only), intra and
+    extracellular water, with a diffusion coefficient D. In this model the
+    echo attenuation of a signal in a single voxel can be written as
+
+        .. math::
+
+        S(b) = S_0[f*e^{(-b*D\*)} + (1-f)e^{(-b*D)}]
+
+        Where:
+        .. math::
+
+        S_0, f, D\* and D are the IVIM parameters.
+
+    Parameters
+    ----------
+    params : array
+             An array of IVIM parameters - S0, f, D_star, D
+
+    bvals  : array
+             bvalues
+
+    References
+    ----------
+    .. [1] Le Bihan, Denis, et al. "Separation of diffusion
+               and perfusion in intravoxel incoherent motion MR
+               imaging." Radiology 168.2 (1988): 497-505.
+    .. [2] Federau, Christian, et al. "Quantitative measurement
+               of brain perfusion with intravoxel incoherent motion
+               MR imaging." Radiology 265.3 (2012): 874-881.
+    """
+    S0, f, D_star, D = params
+    S = S0 * (f * np.exp(-bvals * D_star) + (1 - f) * np.exp(-bvals * D))
+    return S
+
+
+def _ivim_error(params, bvals, signal):
+    """Error function to be used in fitting the IVIM model
+    """
+    return (signal - ivim_function(params, bvals))
+
+
+class IvimModel(ReconstModel):
+    """Ivim model
+    """
+
+    def __init__(self, gtab, split_b=200.0, min_signal=None):
+        """
+        Initialize an IVIM model.
+
+        Parameters
+        ----------
+        gtab : GradientTable class instance
+            Gradient directions and bvalues
+
+        split_b : float
+            The b-value to split the data on for two-stage fit
+
+        min_signal : float
+            The minimum signal value in the data.
+        """
+        ReconstModel.__init__(self, gtab)
+        self.split_b = split_b
+        self.min_signal = min_signal
+        if self.min_signal is not None and self.min_signal <= 0:
+            e_s = "The `min_signal` key-word argument needs to be strictly"
+            e_s += " positive."
+            raise ValueError(e_s)
+
+    def fit(self, data, mask=None, x0=None, fit_method="one_stage",
+            jac=True, bounds=([0., 0., 0., 0.], [np.inf, 1., 1., 1.]),
```

Comment:
These should all be passed to the `__init__` of the `IvimModel` class, and kept as attributes of the model class instance